ast/util.go

```diff
@@ -0,0 +1,91 @@
+package ast
+
+import (
+	"github.com/cloudspannerecosystem/memefish/token"
+	"strings"
+)
+
+// Helper functions for SQL()
+
+// sqlOpt outputs:
+//
+//	when node != nil: left + node.SQL() + right
+//	else            : empty string
+//
+// This function corresponds to sqlOpt in ast.go
+func sqlOpt[T interface {
+	Node
+	comparable
+}](left string, node T, right string) string {
+	var zero T
+	if node == zero {
+		return ""
+	}
+	return left + node.SQL() + right
+}
+
+// strOpt outputs:
+//
+//	when pred == true: s
+//	else            : empty string
+//
+// This function corresponds to {{if pred}}s{{end}} in ast.go
+func strOpt(pred bool, s string) string {
+	if pred {
+		return s
+	}
+	return ""
+}
+
+// sqlJoin outputs joined string of SQL() of all elems by sep.
+// This function corresponds to sqlJoin in ast.go
+func sqlJoin[T Node](elems []T, sep string) string {
+	var b strings.Builder
+	for i, r := range elems {
+		if i > 0 {
+			b.WriteString(sep)
+		}
+		b.WriteString(r.SQL())
+	}
+	return b.String()
+}
+
+// Helper functions for Pos(), End()
+
+// lastElem returns last element of slice s.
```

Comment:
I understand your intention, but this function seems too generic. I would not want to add this as it could be used in places it is not supposed to be.

Reply:
OK, `lastElem` can be written as inline, so I will remove it.

(It may be possible to be accepted if `T` is restricted to `Node`)
```diff
-func lastElem[T any](s []T) T {
+func lastElem[T Node](s []T) T {
```

Reply:
Fixed https://github.com/cloudspannerecosystem/memefish/pull/120/commits/e1a3e815ecd3fca895317482ec3d072e999d1eb5

Reply:
<code>func last<del>Elem</del><ins>Node</ins>[T Node](s []T) T {</code> is acceptable.

Reply:
Introduce `func lastNode[T Node](s []T) T` in https://github.com/cloudspannerecosystem/memefish/pull/120/commits/47ea7094bc4ac09ecce9ef99bb7ce995f743d0e5